source/extensions/common/wasm/wasm.h

```diff
@@ -343,20 +387,41 @@ class Wasm : public Envoy::Server::Wasm,
 
 private:
   friend class Context;
+  static const uint32_t kMetricTypeMask = 0x3;
+  static const uint32_t kMetricTypeCounter = 0x0;
+  static const uint32_t kMetricTypeGauge = 0x1;
+  static const uint32_t kMetricTypeHistogram = 0x2;
+  static const uint32_t kMetricIdIncrement = 0x4;
+
+  bool isCounterMetricId(uint32_t metric_id) {
+    return (metric_id & kMetricTypeMask) == kMetricTypeCounter;
+  }
+  bool isGaugeMetricId(uint32_t metric_id) {
+    return (metric_id & kMetricTypeMask) == kMetricTypeGauge;
+  }
+  bool isHistogramMetricId(uint32_t metric_id) {
+    return (metric_id & kMetricTypeMask) == kMetricTypeHistogram;
+  }
+  uint32_t nextCounterMetricId() { return next_counter_metric_id_ += kMetricIdIncrement; }
+  uint32_t nextGaugeMetricId() { return next_gauge_metric_id_ += kMetricIdIncrement; }
+  uint32_t nextHistogramMetricId() { return next_histogram_metric_id_ += kMetricIdIncrement; }
 
   void registerCallbacks();    // Register functions called out from WASM.
   void establishEnvironment(); // Language specific enviroments.
   void getFunctions();         // Get functions call into WASM.
 
   Upstream::ClusterManager& cluster_manager_;
   Event::Dispatcher& dispatcher_;
+  Stats::Scope& scope_;
   std::string id_;
```

Comment:
Question: the `scope_` is either reference to higher level, or to `owner_scope_`, correct?

Reply:
Correct, commented.